### DISCUSSION TITLE
Can you share tasks efficiently across multiple git repos?

### DISCUSSION THREAD

**Discussion by fredrikaverpil** [go-task/task#1420] [Q&A] [answered]
Let's say you have a bunch of git repos in your organization. In each, you can add a `Taskfile.yml` with commands for the individual project. But let's say that for all these repos there is a common task that ideally would be defined in one central place (not in each of all the repos). 

Is it possible to set up git repo for such shared tasks and somehow have them all available when running `task --list` along with any project-specific tasks in the current project directory?

I saw there are [global tasks](https://taskfile.dev/usage/#running-a-global-taskfile), but this assumes you place the shared `Taskfile.yml` on each developer machine which is not ideal. But this _could_ perhaps work if this local/global `Taskfile.yml` includes a `update-shared-tasks` task which pulls down all tasks from the central git repo. It doesn't seem like it would automatically bring in any tasks from a `Taskfile.yml` in the current working directory as well though.

Any input or ideas on this problem is much appreciated!

EDIT: it looks like something like this is work in progress over [here](https://github.com/go-task/task/issues/1317)?

**Comment by andreynering** [comment, marked as answer]
Hi @fredrikaverpil,

Yes, remote Taskfiles is what you're looking for. For now it is an "experiment", which means you need to enable it with an ENV and it is subject to breaking changes until launched for real.

Given you're interested, it'd be awesome if you gave it try and added your feedback on #1317.

https://taskfile.dev/experiments/remote-taskfiles/

**Reply by fredrikaverpil** [reply]
Hi @andreynering and thanks for creating this awesome project!

I'm currently evaluating Task and this remote Taskfile experiment. I got stuck at trying to access a Taskfile sitting in a private GitHub repo. Do you have any tips for me on how to solve this?
I could for example use `$GOPRIVATE` and `$GITHUB_TOKEN`, I'm thinking. But not sure if this is at all supported by the experiment right now, and how to define these with the `includes:` directive - or if there is some other recommended way?

```
includes:
  private: https://raw.githubusercontent.com/myorg/poc-tasks/master/Taskfile.yml
```

```bash
$ task --list-all
task: Download of "https://raw.githubusercontent.com/myorg/poc-tasks/master/Taskfile.yml" failed with status code
 404 (Not Found)
```

Would you rather have me post about this in #1317?

**Reply by pd93** [reply]
Hi @fredrikaverpil, we plan to add support for accessing Taskfiles via the `git`/`ssh` protocol in the future. This will allow you to access private repositories. However, this is not currently implemented in the experiment.

To my knowledge the only way to access private content on GitHub via HTTPS is by adding a PAT in the request header. This is not something that Task supports and I'd prefer to keep it this way as I don't want to encourage people to store secrets in a Taskfile. Unfortunately, I think this leaves you out of options for now unless I've missed a workaround.

**Reply by fredrikaverpil** [reply]
Hi @pd93 😄 

I'm noticing that inserting the actual token works:

```yml
includes:
  private: https://git:aCtUalToKeNHerE@raw.githubusercontent.com/myorg/poc-tasks/master/Taskfile.yml
```

But I would rather that it would read an environment variable, like `$GITHUB_TOKEN` from the environment, like this:


```yml
includes:
  private: https://git:$GITHUB_TOKEN@raw.githubusercontent.com/myorg/poc-tasks/master/Taskfile.yml
```

I'm evaluating using Task as task runner at work, but not being able to use remote Taskfiles in private repos is a total blocker, unfortunately. It would be great if Task could provide some interim solution that would be more of a workaround than the final solution.

**Reply by andreynering** [reply]
@pd93 I actually think that supporting ENVs like `TASK_GITHUB_TOKEN` and `TASK_GITLAB_TOKEN`, etc would be a great idea.

It'd allow us to support private repos in a simple way without having to implement Git support necessarily. And it's secure, because each user would have to set this token on their own machine.

What do you think?

**Reply by pd93** [reply]
I'm not against supporting environment variables. It's definitely easier/quicker to implement than git/ssh. Though also less flexible. I think either way, we will have to support git/ssh in the long run. For example, some private repos have https disabled entirely. Do we want different env vars per upstream or a single, generic one and domain detection?

**Reply by fredrikaverpil** [reply]
> It'd allow us to support private repos in a simple way without having to implement Git support necessarily.

I would say there's likely a demand for ssh too. But starting out by reading a $TOKEN environment variable would be quicker to implement, like you say, and enable some people.

Depending on how this is implemented, git's [insteadOf](https://git-scm.com/docs/git-config#Documentation/git-config.txt-urlltbasegtinsteadOf) could perhaps be used to enable ssh use. But I'm not really sure if it would cover the true ssh story.

```bash
# warning, this changes the local git config

# use https for git
git config --global url."https://github.com/".insteadOf git@github.com:
git config --global url."https://".insteadOf git://

# use git for https
git config --global url."git@github.com:".insteadOf https://github.com/
git config --global url."git://".insteadOf https://
```

> And it's secure, because each user would have to set this token on their own machine.

You also have the use case where the Taskfile is included from a private repo and the tasks need to run in e.g. GitHub Actions. In this case, I'm not really sure what the best way would be for Task (and the `includes` declaration) to understand the secret.

**Reply by andreynering** [reply]
I think both makes sense, indeed. I'd go for ENVs over HTTPS first, and consider adding support for SSH afterwards.

And just a reminder: Go has a native Git implementation, which is probably the way to go here, to avoid depending on local tools. Needs to check if they have proper support for the SSH keys the user has alredy setup, of course.

https://github.com/go-git/go-git

**Reply by fredrikaverpil** [reply]
Here's my full use case:

### Personal tasks

- I'm using `~/Taskfile.yml` for personal needs: https://github.com/fredrikaverpil/dotfiles/blob/main/Taskfile.yml
- I'm cloning down a private github repo of mine to local disk, in which I store sensitive tasks.
- This use case does not need any authentication for the `includes` directive as I can include from a local path.

### Tasks at my company

- Tasks are defined in private git repos.
- Some git repos might benefit from sharing tasks. In this case, authentication is required for the `includes` directive.
- Tasks needs to run locally, which means e.g. defining the `$GITHUB_TOKEN` in the `includes` directive.
- Tasks needs to run in GitHub Actions. In this case, GitHub Actions needs to somehow pass [secrets](https://docs.github.com/en/actions/security-guides/using-secrets-in-github-actions) (`{{ secrets.MY_SECRET }}`) to the `task` command such that the `includes` directive can pick up on it and use it.
- Git repos may be including tasks from other repos in a nested fashion. Nesting seems to work out of the box, but again here authentication is required for the `includes` directive.
